torchvision/csrc/ops/quantized/cpu/qroi_align_kernel.cpp

```diff
@@ -0,0 +1,210 @@
+#include <ATen/ATen.h>
+#include <ATen/native/quantized/affine_quantizer.h>
+#include <torch/library.h>
+
+#include "../../roi_align.h"
+
+namespace vision {
+namespace ops {
+
+namespace {
+
+template <typename T>
+void qroi_align_forward_kernel_impl(
+    int n_rois,
+    const at::Tensor& t_input,
+    const float& spatial_scale,
+    int channels,
+    int height,
+    int width,
+    int pooled_height,
+    int pooled_width,
+    int sampling_ratio,
+    bool aligned,
+    const at::Tensor& t_rois,
+    T* output) {
+  const T* input = t_input.contiguous().data_ptr<T>();
+  int64_t input_zp = t_input.q_zero_point();
+  float input_scale = t_input.q_scale();
+
+  const T* rois = t_rois.contiguous().data_ptr<T>();
+  int64_t rois_zp = t_rois.q_zero_point();
+  float rois_scale = t_rois.q_scale();
+
+  for (int n = 0; n < n_rois; n++) {
+    int index_n = n * channels * pooled_width * pooled_height;
+
+    const T* offset_rois = rois + n * 5;
+    int roi_batch_ind = at::native::dequantize_val(
+        rois_scale, rois_zp, offset_rois[0]);
```

Comment:
I think we might have issues here with large images and batches.

Imagine that the image is of size 1000x1000, and the boxes thus can be from 0 to 1000 in value. The quantized boxes (which have the same quantizers for all boxes) would not be able to properly index the 0, 1, 2, etc values needed for the image index in the batch.

I think the solution for this is either
- [easy] disable batch support altogether in the quantized version of the op
- [medium] try using per-channel quantization of the boxes (so that the quantizer for the indices is different than the quantizer for the coordinates) (need to see how it integrates with the rest of the pipeline)
- [harder] break BC in the kernels and pass instead a list of tensors for the rois. This would maybe involve more changes down the road in the Faster R-CNN model and would probably be tricky, so I wouldn't recommend this.

Additionally, can you also add a quick test exercising this code-path that I mentioned (with a large image and a few batch elements)?

Reply:
Yes, if I understand your concern correctly, this is similar to my previous https://github.com/pytorch/vision/pull/3624#discussion_r607027926:

> we can't represent 1 with (scale, zero_point) = (2, 0)

I pushed a check in d6f78ab0af9ef3f859733645191b5ebfb57dd1ba as we discussed yesterday, but I realize now that it's not correct/enough anyway: it assumes that (scale, zero_point) == (1, 0) which is wrong in general.